scripts-base/src/Component.ts

```diff
@@ -0,0 +1,96 @@
+import {
+	ComponentEl,
+	DelegateEvent,
+	DelegateTarget,
+	EventListeners,
+	EventMap,
+	EventMapByElement,
+} from './componentTypes'
+import { matchesSelector } from './utils'
+
+export interface NamedComponent {
+	componentName: string
+}
+
+export type ComponentConstructor<
+	D,
+	E extends ComponentEl,
+	EMap extends EventMap = EventMapByElement<E>
+> = NamedComponent & (new (element: E, data: D) => Component<D, E, EMap>)
+
+export class ComponentInitializationError extends Error {}
+
+export class Component<D = {}, E extends ComponentEl = HTMLElement, EMap extends EventMap = EventMapByElement<E>> {
+	protected readonly getListeners = (): EventListeners<E, EMap> => []
+
+	constructor(protected readonly el: E, protected readonly data: D) {}
+
+	public setup() {
+		this.attachListeners()
+
+		this.init()
+	}
+
+	protected getChild<C extends DelegateTarget<E>>(
+		selector: string,
+		ChildConstructor: new (...args: any[]) => C,
+		parent: HTMLElement | SVGElement = this.el as HTMLElement | SVGElement
+	): C {
+		const child = parent.querySelector(selector)
+
+		if (!(child instanceof ChildConstructor)) {
+			throw new ComponentInitializationError(
```

Comment:
Pojmenování `ComponentInitializationError` a obsah textu chybu mi významově spolu vůbec nesouvisí. → `InvalidChildTypeError` bych předpokládal podle textu. Initialization fáze komponenty je v době tohoto volání už dávno pryč.

Reply:
To vůbec není pravda. Viz předchozí komentář.
Navíc kdyby komponenta mohla vyhazovat nekonečno výjimek, muselo by být `initializeComponents` zbytečně složité a větší.

Reply:
`this.getChild()` můžu volat v během celého lifecycle komponenty. Když mi to během práce s komponentou vyhodí `ComponentInitializationError`, bude mne to zbytečně mást, místo abych věděl rovnou o co jde (proto navrhovaný název `InvalidChildTypeError`)

Reply:
K tomu to ale není určené. Je to zkratka pro _assertion_, že ten potomek existuje a má daný typ, abys to nemusel řešit a ifovat sám. Navíc když to zavoláš mimo `constructor` nebo `init`, `initializeComponents` Ti to nechytí, tak to ani nemáš dělat. Je to pro *essential* children, což jsou v praxi takřka všechny.